x-pack/auditbeat/tracing/perfevent.go

```diff
@@ -0,0 +1,509 @@
+	return func(channel *PerfChannel) error {
+		channel.pollTimeout = timeout
+		return nil
+	}
+}
+
+// MonitorProbe associates a probe with the PerfChannel, so that events
+// generated by this probe will be received. A probe is identified by its
+// ProbeFormat. The Decoder is used to decode events from this probe and
+// will determine the types and contents of the returned events.
+func (c *PerfChannel) MonitorProbe(format ProbeFormat, decoder Decoder) error {
+	c.attr.Config = uint64(format.ID)
+	doGroup := len(c.evs) > 0
+	for idx := 0; idx < c.numCPUs; idx++ {
+		var group *perf.Event
+		var flags int
+		if doGroup {
+			group = c.evs[idx]
+			flags = unix.PERF_FLAG_FD_NO_GROUP | unix.PERF_FLAG_FD_OUTPUT
+		}
+		ev, err := perf.OpenWithFlags(&c.attr, c.pid, idx, group, flags)
+		if err != nil {
+			return err
+		}
+		cid, err := ev.ID()
+		if err != nil {
+			return err
+		}
+		if len(format.Probe.Filter) > 0 {
+			fd, err := ev.FD()
+			if err != nil {
+				return err
+			}
+			fbytes := []byte(format.Probe.Filter + "\x00")
+			_, _, errNo := unix.Syscall(unix.SYS_IOCTL, uintptr(fd), unix.PERF_EVENT_IOC_SET_FILTER, uintptr(unsafe.Pointer(&fbytes[0])))
+			if errNo != 0 {
+				return errors.Wrapf(errNo, "unable to set filter '%s'", format.Probe.Filter)
+			}
+		}
+		c.streams[cid] = stream{probeID: format.ID, decoder: decoder}
+		c.evs = append(c.evs, ev)
+
+		if !doGroup {
+			if err := ev.MapRingNumPages(c.mappedPages); err != nil {
+				return errors.Wrap(err, "perf channel mapring failed")
+			}
+		}
+	}
+	return nil
+}
+
+// C returns the channel to read samples from.
+func (c *PerfChannel) C() <-chan interface{} {
+	return c.sampleC
+}
+
+// ErrC returns the channel to read errors from.
+func (c *PerfChannel) ErrC() <-chan error {
+	return c.errC
+}
+
+// LostC returns the channel to read lost samples notifications.
+func (c *PerfChannel) LostC() <-chan uint64 {
+	return c.lostC
+}
+
+// Run enables the configured probe and starts receiving perf events.
+// sampleC is the channel where decoded perf events are received.
+// errC is the channel where errors are received.
+//
+// The format of the received events depends on the Decoder used.
+func (c *PerfChannel) Run() error {
+	if !atomic.CompareAndSwapUintptr(&c.running, 0, 1) {
+		return ErrAlreadyRunning
+	}
+	c.sampleC = make(chan interface{}, c.sizeSampleC)
+	c.errC = make(chan error, c.sizeErrC)
+	c.lostC = make(chan uint64, c.sizeLostC)
+
+	for _, ev := range c.evs {
+		if err := ev.Enable(); err != nil {
+			return errors.Wrap(err, "perf channel enable failed")
+		}
+	}
+	c.wg.Add(1)
+	go c.channelLoop()
+	return nil
+}
+
+// Close closes the channel.
+func (c *PerfChannel) Close() error {
+	if atomic.CompareAndSwapUintptr(&c.running, 1, 2) {
+		close(c.done)
+		c.wg.Wait()
+		defer close(c.sampleC)
+		defer close(c.errC)
+		defer close(c.lostC)
+	}
+	var errs multierror.Errors
+	for _, ev := range c.evs {
+		if err := ev.Disable(); err != nil {
+			errs = append(errs, err)
```

Comment:
Add context to these errors with `errors.Wrap`?

Reply:
done